topi/python/topi/image/resize.py

```diff
@@ -210,3 +210,185 @@ def _bicubic(*indices):
+        output_shape = [box_indices.shape[0], data.shape[1],
+                        crop_size[0], crop_size[1], data.shape[4]]
+        image_height = data.shape[2]
+        image_width = data.shape[3]
+
+    def _get_pixel(n, c, y, x, cc):
+        if layout.lower() == 'nhwc':
+            return data(n, y.astype("int32"), x.astype("int32"), c).astype('float')
+        if layout.lower() == 'nchw':
+            return data(n, c, y.astype("int32"), x.astype("int32")).astype('float')
+        # else must be NCHWxc
+        return data(n, c, y.astype("int32"), x.astype("int32"), cc).astype('float')
+
+    def _get_indices(*indices):
+        if layout == 'NHWC':
+            n, y, x, c = indices
+            cc = None
+        elif layout == 'NCHW':
+            n, c, y, x = indices
+            cc = None
+        else:
+            n, c, y, x, cc = indices
+
+        return n, c, y, x, cc
+
+    def _cast_output(value):
+        if out_dtype:
+            dtype = out_dtype
+        else:
+            dtype = data.dtype
+        return value.astype(dtype)
+
+    # Nearest neighbor computation
+    def _nearest_neighbor(*indices):
+        n, c, y, x, cc = _get_indices(*indices)
+        box_idx = box_indices(n)
+
+        y1, x1 = boxes(n, 0), boxes(n, 1)
+        y2, x2 = boxes(n, 2), boxes(n, 3)
+
+        in_h = (image_height - 1) * (y2 - y1)
+        in_w = (image_width - 1) * (x2 - x1)
+        h_scale = tvm.div(in_h, target_h - 1)
+        w_scale = tvm.div(in_w, target_w - 1)
+
+        in_y = y1 * (image_height - 1) + h_scale * y
+        in_x = x1 * (image_width - 1) + w_scale * x
+        closest_x_index = tvm.round(in_x)
+        closest_y_index = tvm.round(in_y)
+
+        value = _get_pixel(box_idx, c, closest_y_index, closest_x_index, cc)
+        out_y = tvm.if_then_else(in_y < 0,
+                                 extrapolation_value,
+                                 tvm.if_then_else(in_y > image_height - 1,
+                                                  extrapolation_value,
+                                                  value))
+
+        # use extrapolation_value if in_x is out of boundary
+        out = tvm.if_then_else(in_x < 0,
+                               extrapolation_value,
+                               tvm.if_then_else(in_x > image_width - 1,
+                                                extrapolation_value,
+                                                out_y))
+        return _cast_output(out)
+
+
+    # Bilinear helper functions and computation.
+    def _lerp(A, B, t):
+        return A * (1.0 - t) + B * t
+
+    def _bilinear(*indices):
+        n, c, y, x, cc = _get_indices(*indices)
+        box_idx = box_indices(n)
+
+        y1, x1 = boxes(n, 0), boxes(n, 1)
+        y2, x2 = boxes(n, 2), boxes(n, 3)
+
+        in_h = (image_height - 1) * (y2 - y1)
+        in_w = (image_width - 1) * (x2 - x1)
+        h_scale = tvm.div(in_h, target_h - 1)
+        w_scale = tvm.div(in_w, target_w - 1)
```

Comment:
Agree with @jwfromm This seeems to be repeated.

Reply:
It is sligthtly different from crop, for example, h_scale is calculated based on boxes while crop has no boxes, etc. I have moved these kernels out of the op and generalized them to work for both resize and crop_and_resize.